spec/client/ics-006-solo-machine-client/README.md

```diff
@@ -127,205 +147,178 @@ function latestClientHeight(clientState: ClientState): uint64 {
-  clientState.consensusState.publicKey = header.newPublicKey
-  clientState.consensusState.diversifier = header.newDiversifier
-  clientState.consensusState.timestamp = header.timestamp
-  clientState.consensusState.sequence++
+function  verifyClientMessage(clientMsg: ClientMessage) {
+  switch typeof(ClientMessage) {
+    case Header:
+      verifyHeader(clientMessage)
+    // misbehaviour only suppported for current public key and diversifier on solomachine
+    case Misbehaviour:
+      verifyMisbehaviour(clientMessage)
+  }
 }
-```
-
-### Misbehaviour predicate
 
-Any duplicate signature on different messages by the current public key freezes a solo machine client.
+function verifyHeader(header: header) {
+    assert(header.timestamp >= clientstate.consensusState.timestamp)
+    headerData = {
+      NewPublicKey: header.newPublicKey,
+      NewDiversifier: header.newDiversifier,
+    }
+    sigBytes = SignBytes(
+      Sequence: clientState.consensusState.sequence,
+      Timestamp: header.timestamp,
+      Diversifier: clientState.consensusState.diversifier,
+      Path: []byte{"solomachine:header"}
+      Value: marshal(headerData)
+    )
+    assert(checkSignature(cs.consensusState.publicKey, sigBytes, header.signature))
+}
 
-```typescript
-function checkMisbehaviourAndUpdateState(
-  clientState: ClientState,
-  misbehaviour: Misbehaviour) {
-    h1 = misbehaviour.h1
-    h2 = misbehaviour.h2
+function verifyMisbehaviour(misbehaviour: Misbehaviour) {
+    s1 = misbehaviour.signatureOne
+    s2 = misbehaviour.signatureTwo
     pubkey = clientState.consensusState.publicKey
     diversifier = clientState.consensusState.diversifier
     timestamp = clientState.consensusState.timestamp
     // assert that timestamp could have fooled the light client
-    assert(misbehaviour.h1.signature.timestamp >= timestamp)
-    assert(misbehaviour.h2.signature.timestamp >= timestamp)
-    // assert that signature data is different
-    assert(misbehaviour.h1.signature.data !== misbehaviour.h2.signature.data)
-    // assert that the signatures validate
-    assert(checkSignature(pubkey, misbehaviour.sequence, diversifier, misbehaviour.h1.signature.data))
-    assert(checkSignature(pubkey, misbehaviour.sequence, diversifier, misbehaviour.h2.signature.data))
-    // freeze the client
-    clientState.frozen = true
+    assert(misbehaviour.s1.timestamp >= timestamp)
+    assert(misbehaviour.s2.timestamp >= timestamp)
+    // assert that the signatures validate and that they are different
+    sigBytes1 = SignBytes(
+      Sequence: misbehaviour.sequence,
+      Timestamp: s1.timestamp,
+      Diversifier: diversifier,
+      Path: s1.path,
+      Data: s1.data
+    )
+    sigBytes2 = SignBytes(
+      Sequence: misbehaviour.sequence,
+      Timestamp: s2.timestamp,
+      Diversifier: diversifier,
+      Path: s2.path,
+      Data: s2.data
+    )
+    assert(sigBytes1 != sigBytes2)
```

Comment:
Okay so, in ibc-go we should remove [this check](https://github.com/cosmos/ibc-go/blob/main/modules/light-clients/06-solomachine/misbehaviour.go#L44). That is, if all the fields are identical except the timestamps, that should be considered misbehaviour?

Reply:
No the implementation is correct. good catch